CppCoreGuidelines.md

```diff
@@ -5968,26 +5969,30 @@ You can handle self-assignment by explicitly testing for self-assignment, but of
 
     Foo& Foo::operator=(const Foo& a)   // OK, but there is a cost
     {
-        if (this == &a) return *this;
-        s = a.s;
-        i = a.i;
+        if (this != &a) {
+            s = a.s;
+            i = a.i;
+        }
         return *this;
     }
 
-This is obviously safe and apparently efficient.
-However, what if we do one self-assignment per million assignments?
-That's about a million redundant tests (but since the answer is essentially always the same, the computer's branch predictor will guess right essentially every time).
-Consider:
+The code we are skipping with the test would actually work fine even for self-assignment;
+so the extra test is not buying us any correctness, only speed. And whatever speed we gain
+in the rare case of self-assignment may well be lost again, in that same case, due to
+branch misprediction.
+
+Furthermore, while the code above is technically safe, careless maintenance
+(such as swapping the order of members `s` and `i` or changing the type of `i`) may cause it
+to lose the strong exception guarantee. To preserve correctness, we would do better
+to use the "copy and swap" idiom:
```

Comment:
So you're suggesting to replace the almost invisible costs of a single comparison with an entire copy-and-swap?

Reply:
Replace the almost invisible costs of the _entire function_ with the almost invisible costs of a copy-and-swap.

It's true that `string::operator=` can be faster than `string::string(const string&)`, because the latter must do a heap allocation _and_ a memcpy into the resulting buffer, whereas the former can skip the heap allocation if its buffer is already big enough (but still pays for the memcpy).

QuickBench can be massaged to show various results. The smaller the string is, the better simple-assignment performs. Copy-and-swap beats simple-assignment on libc++; I don't really see why. http://quick-bench.com/YiLlbTxctQKuvRINZVkg1h4MKWw

As far as the C++ Core Guidelines are concerned, though, we're not talking about performance micro-optimizations; we're talking about guidelines that help programmers avoid big correctness issues like

    Foo& Foo::operator=(const Foo& a) {
        i = a.i;
        s = a.s;    // ERROR: This code is not exception-safe!
        return *this;
    }

The old text went so far as to claim that there was "no known general way of avoiding an `if (this == &a) return *this;` test for a move assignment and still get a correct answer" — as if copy-and-swap hadn't been known for decades!